src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java

```diff
@@ -114,40 +136,142 @@ static SqlAuthenticationToken getSqlFedAuthTokenPrincipal(SqlFedAuthInfo fedAuth
+
         try {
-            /*
-             * principal name does not matter, what matters is the realm name it gets the username in
-             * principal_name@realm_name format
-             */
-            KerberosPrincipal kerberosPrincipal = new KerberosPrincipal("username");
-            String user = kerberosPrincipal.getName();
+            ConfidentialClientApplication clientApplication = null;
+
+            // check if cert is PKCS12 first
+            try (InputStream is = new FileInputStream(certFile)) {
+                KeyStore keyStore = SQLServerCertificateUtils.loadPKCS12KeyStore(certFile, certPassword);
+
+                if (logger.isLoggable(Level.FINEST)) {
+                    logger.finest(logContext + "certificate type: " + keyStore.getType());
+
+                    // we don't need to do this unless logging enabled since MSAL will fail if cert is not valid
+                    Enumeration<String> enumeration = keyStore.aliases();
+                    while (enumeration.hasMoreElements()) {
+                        String alias = (String) enumeration.nextElement();
+                        X509Certificate cert = (X509Certificate) keyStore.getCertificate(alias);
+                        cert.checkValidity();
+                        logger.finest(logContext + "certificate: " + cert.toString());
+                    }
+                }
+
+                IClientCredential credential = ClientCredentialFactory.createFromCertificate(is, certPassword);
+                clientApplication = ConfidentialClientApplication.builder(aadPrincipalID, credential)
+                        .executorService(executorService)
+                        .setTokenCacheAccessAspect(PersistentTokenCacheAccessAspect.getInstance())
+                        .authority(fedAuthInfo.stsurl).build();
+            } catch (FileNotFoundException e) {
+                // re-throw if file not there no point to try another format
+                throw e;
+            } catch (CertificateException | NoSuchAlgorithmException | IOException e) {
+                // ignore not PKCS12 cert error, will try another format after this
+                if (logger.isLoggable(Level.FINEST)) {
+                    logger.finest(logContext + "Error loading PKCS12 certificate: " + e.getMessage());
+                }
+            }
+
+            if (clientApplication == null) {
+                // try loading X509 cert
+                X509Certificate cert = (X509Certificate) SQLServerCertificateUtils.loadCertificate(certFile);
+
+                if (logger.isLoggable(Level.FINEST)) {
+                    logger.finest(logContext + "certificate type: " + cert.getType());
+
+                    // we don't really need to do this, MSAL will fail if cert is not valid, but good to check here and throw with proper error message
+                    cert.checkValidity();
+                    logger.finest(logContext + "certificate: " + cert.toString());
```

Comment:
Unnecessary toString().
```suggestion
                    logger.finest(logContext + "certificate: " + cert);
```